### Задание 1

Импортируйте библиотеки pandas и numpy.

Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. Создайте датафреймы X и y из этих данных.

Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.

Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.

Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.

Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.

In [1]:
import numpy as np
import pandas as pd

In [2]:
from sklearn.datasets import load_boston
boston = load_boston()
X = pd.DataFrame(boston.data, columns = boston.feature_names)
y = pd.DataFrame(boston.target, columns = ['price'])

In [3]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 42)

In [4]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()

In [5]:
lr.fit(X_train, y_train)
y_pred = lr.predict(X_test)

In [6]:
from sklearn.metrics import r2_score
R2 = r2_score(y_test, y_pred)
R2

0.711226005748496

### Задание 2

Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble.

Сделайте агрумент n_estimators равным 1000, max_depth должен быть равен 12 и random_state сделайте равным 42.

Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression, но при этом в метод fit вместо датафрейма y_train поставьте `y_train.values[:, 0]`, чтобы получить из датафрейма одномерный массив Numpy, так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.

Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания.

Напишите в комментариях к коду, какая модель в данном случае работает лучше.

In [7]:
from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor(n_estimators = 1000, max_depth = 12, random_state = 42)
model

RandomForestRegressor(max_depth=12, n_estimators=1000, random_state=42)

In [8]:
model.fit(X_train, y_train.values[:, 0])

RandomForestRegressor(max_depth=12, n_estimators=1000, random_state=42)

In [9]:
y_pred = model.predict(X_test)
R2 = r2_score(y_test, y_pred)
R2

0.87472606157312

При использовании LinearRegression результат R2 = 0.711226005748496, а при использовании RandomForestRegressor значение R2 = 0.87472606157312. Отсюда видно, что модель, построенная с помощью RandomForestRegressor работает лучше, чем LinearRegression, на датасете sklearn.datasets.load_boston

### *Задание 3

Вызовите документацию для класса RandomForestRegressor, найдите информацию об атрибуте feature_importances_.

С помощью этого атрибута найдите сумму всех показателей важности, установите, какие два признака показывают наибольшую важность.

In [10]:
?RandomForestRegressor

In [11]:
fi = model.feature_importances_
fi

array([0.03167574, 0.00154252, 0.00713813, 0.00123624, 0.01426897,
       0.40268179, 0.01429864, 0.06397257, 0.00528122, 0.01152493,
       0.01808108, 0.01245085, 0.41584732])

In [12]:
np.sum(fi)

1.0

In [13]:
ind = np.argpartition(fi, -2)[-2:]
ind

array([ 5, 12], dtype=int64)

In [14]:
X_train.columns

Index(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX',
       'PTRATIO', 'B', 'LSTAT'],
      dtype='object')

In [15]:
X_train.columns[ind]

Index(['RM', 'LSTAT'], dtype='object')

### *Задание 4

В этом задании мы будем работать с датасетом, с которым мы уже знакомы по домашнему заданию по библиотеке Matplotlib, это датасет Credit Card Fraud Detection. Для этого датасета мы будем решать задачу классификации - будем определять, какие из транзакциции по кредитной карте являются мошенническими. Данный датасет сильно несбалансирован (так как случаи мошенничества относительно редки), так что применение метрики accuracy не принесет пользы и не поможет выбрать лучшую модель. Мы будем вычислять AUC, то есть площадь под кривой ROC.

Импортируйте из соответствующих модулей RandomForestClassifier, GridSearchCV и train_test_split.

Загрузите датасет creditcard.csv и создайте датафрейм df.

С помощью метода value_counts с аргументом `normalize=True` убедитесь в том, что выборка несбалансирована. Используя метод info, проверьте, все ли столбцы содержат числовые данные и нет ли в них пропусков. Примените следующую настройку, чтобы можно было просматривать все столбцы датафрейма: `pd.options.display.max_columns = 100`.

Просмотрите первые 10 строк датафрейма df.

Создайте датафрейм X из датафрейма df, исключив столбец Class.

Создайте объект Series под названием y из столбца Class.

Разбейте X и y на тренировочный и тестовый наборы данных при помощи функции train_test_split, используя аргументы: `test_size=0.3, random_state=100, stratify=y`. У вас должны получиться объекты X_train, X_test, y_train и y_test.

Просмотрите информацию о их форме.

Для поиска по сетке параметров задайте такие параметры: `parameters = [{'n_estimators': [10, 15]`, `'max_features': np.arange(3, 5)`, `'max_depth': np.arange(4, 7)}]`.

Создайте модель GridSearchCV со следующими аргументами: `estimator=RandomForestClassifier(random_state=100)`, `param_grid=parameters`, `scoring='roc_auc'`, `cv=3`.

Обучите модель на тренировочном наборе данных (может занять несколько минут).

Просмотрите параметры лучшей модели с помощью атрибута best_params_.

Предскажите вероятности классов с помощью полученнной модели и метода predict_proba.

Из полученного результата (массив Numpy) выберите столбец с индексом 1 (вероятность класса 1) и запишите в массив y_pred_proba. Из модуля sklearn.metrics импортируйте метрику roc_auc_score.

Вычислите AUC на тестовых данных и сравните с результатом,полученным на тренировочных данных, используя в качестве аргументов массивы y_test и y_pred_proba.

In [16]:
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier

In [17]:
df = pd.read_csv('creditcard.csv')
df

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[284807 rows x 31 columns]

In [18]:
pd.options.display.max_columns = 100
df.value_counts(normalize=True)

Time      V1          V2          V3         V4         V5         V6         V7          V8          V9         V10        V11        V12        V13        V14        V15        V16        V17        V18        V19        V20        V21         V22        V23        V24        V25        V26        V27        V28        Amount  Class
163152.0  -1.196037    1.585949    2.883976   3.378471   1.511706   3.717077   0.585362   -0.156001    0.122648   4.217934   1.385525  -0.709405  -0.256168  -1.564352   1.693218  -0.785210  -0.228008  -0.412833   0.234834   1.375790  -0.370294    0.524395  -0.355170  -0.869790  -0.133198   0.327804  -0.035702  -0.858197  7.56    0        0.000063
          -1.203617    1.574009    2.889277   3.381404   1.538663   3.698747   0.560211   -0.150911    0.124136   4.220998   1.384569  -0.706897  -0.256274  -1.562583   1.692915  -0.787338  -0.226776  -0.412354   0.234322   1.385597  -0.366727    0.522223  -0.357329  -0.870174  -0.134166   0.327019  -0.042648  -0

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [20]:
df.head(10)

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   
5   2.0 -0.425966  0.960523  1.141109 -0.168252  0.420987 -0.029728  0.476201   
6   4.0  1.229658  0.141004  0.045371  1.202613  0.191881  0.272708 -0.005159   
7   7.0 -0.644269  1.417964  1.074380 -0.492199  0.948934  0.428118  1.120631   
8   7.0 -0.894286  0.286157 -0.113192 -0.271526  2.669599  3.721818  0.370145   
9   9.0 -0.338262  1.119593  1.044367 -0.222187  0.499361 -0.246761  0.651583   

         V8        V9       V10       V11       V12       V13       V14  \
0  0.098698  0.363787  0.090794 -0.551600 -0.617801 -0.991390 -0.311169   
1  0.085102 -0.255425 -0.166974  1.612727  1.065235  0.489095 -0.143772   
2  0.247676 -1.514654  0.207643  0.624501  0.066084  0.717293 -0.165946   
3  0.377436 -1.387024 -0.054952 -0.226487  0.178228  0.507757 -0.287924   
4 -0.270533  0.817739  0.753074 -0.822843  0.538196  1.345852 -1.119670   
5  0.260314 -0.568671 -0.371407  1.341262  0.359894 -0.358091 -0.137134   
6  0.081213  0.464960 -0.099254 -1.416907 -0.153826 -0.751063  0.167372   
7 -3.807864  0.615375  1.249376 -0.619468  0.291474  1.757964 -1.323865   
8  0.851084 -0.392048 -0.410430 -0.705117 -0.110452 -0.286254  0.074355   
9  0.069539 -0.736727 -0.366846  1.017614  0.836390  1.006844 -0.443523   

        V15       V16       V17       V18       V19       V20       V21  \
0  1.468177 -0.470401  0.207971  0.025791  0.403993  0.251412 -0.018307   
1  0.635558  0.463917 -0.114805 -0.183361 -0.145783 -0.069083 -0.225775   
2  2.345865 -2.890083  1.109969 -0.121359 -2.261857  0.524980  0.247998   
3 -0.631418 -1.059647 -0.684093  1.965775 -1.232622 -0.208038 -0.108300   
4  0.175121 -0.451449 -0.237033 -0.038195  0.803487  0.408542 -0.009431   
5  0.517617  0.401726 -0.058133  0.068653 -0.033194  0.084968 -0.208254   
6  0.050144 -0.443587  0.002821 -0.611987 -0.045575 -0.219633 -0.167716   
7  0.686133 -0.076127 -1.222127 -0.358222  0.324505 -0.156742  1.943465   
8 -0.328783 -0.210077 -0.499768  0.118765  0.570328  0.052736 -0.073425   
9  0.150219  0.739453 -0.540980  0.476677  0.451773  0.203711 -0.246914   

        V22       V23       V24       V25       V26       V27       V28  \
0  0.277838 -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053   
1 -0.638672  0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724   
2  0.771679  0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752   
3  0.005274 -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458   
4  0.798278 -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   
5 -0.559825 -0.026398 -0.371427 -0.232794  0.105915  0.253844  0.081080   
6 -0.270710 -0.154104 -0.780055  0.750137 -0.257237  0.034507  0.005168   
7 -1.015455  0.057504 -0.649709 -0.415267 -0.051634 -1.206921 -1.085339   
8 -0.268092 -0.204233  1.011592  0.373205 -0.384157  0.011747  0.142404   
9 -0.633753 -0.120794 -0.385050 -0.069733  0.094199  0.246219  0.083076   

   Amount  Class  
0  149.62      0  
1    2.69      0  
2  378.66      0  
3  123.50      0  
4   69.99      0  
5    3.67      0  
6    4.99      0  
7   40.80      0  
8   93.20      0  
9    3.68      0

In [21]:
X = df.drop(columns = 'Class')
X

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9       V10       V11       V12  \
0       0.462388  0.239599  0.098698  0.363787  0.090794 -0.551600 -0.617801   
1      -0.082361 -0.078803  0.085102 -0.255425 -0.166974  1.612727  1.065235   
2       1.800499  0.791461  0.247676 -1.514654  0.207643  0.624501  0.066084   
3       1.247203  0.237609  0.377436 -1.387024 -0.054952 -0.226487  0.178228   
4       0.095921  0.592941 -0.270533  0.817739  0.753074 -0.822843  0.538196   
...          ...       ...       ...       ...       ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  4.356170 -1.593105  2.711941   
284803  1.058415  0.024330  0.294869  0.584800 -0.975926 -0.150189  0.915802   
284804  3.031260 -0.296827  0.708417  0.432454 -0.484782  0.411614  0.063119   
284805  0.623708 -0.686180  0.679145  0.392087 -0.399126 -1.933849 -0.962886   
284806 -0.649617  1.577006 -0.414650  0.486180 -0.915427 -1.040458 -0.031513   

             V13       V14       V15       V16       V17       V18       V19  \
0      -0.991390 -0.311169  1.468177 -0.470401  0.207971  0.025791  0.403993   
1       0.489095 -0.143772  0.635558  0.463917 -0.114805 -0.183361 -0.145783   
2       0.717293 -0.165946  2.345865 -2.890083  1.109969 -0.121359 -2.261857   
3       0.507757 -0.287924 -0.631418 -1.059647 -0.684093  1.965775 -1.232622   
4       1.345852 -1.119670  0.175121 -0.451449 -0.237033 -0.038195  0.803487   
...          ...       ...       ...       ...       ...       ...       ...   
284802 -0.689256  4.626942 -0.924459  1.107641  1.991691  0.510632 -0.682920   
284803  1.214756 -0.675143  1.164931 -0.711757 -0.025693 -1.221179 -1.545556   
284804 -0.183699 -0.510602  1.329284  0.140716  0.313502  0.395652 -0.577252   
284805 -1.042082  0.449624  1.962563 -0.608577  0.509928  1.113981  2.897849   
284806 -0.188093 -0.084316  0.041333 -0.302620 -0.660377  0.167430 -0.256117   

             V20       V21       V22       V23       V24       V25       V26  \
0       0.251412 -0.018307  0.277838 -0.110474  0.066928  0.128539 -0.189115   
1      -0.069083 -0.225775 -0.638672  0.101288 -0.339846  0.167170  0.125895   
2       0.524980  0.247998  0.771679  0.909412 -0.689281 -0.327642 -0.139097   
3      -0.208038 -0.108300  0.005274 -0.190321 -1.175575  0.647376 -0.221929   
4       0.408542 -0.009431  0.798278 -0.137458  0.141267 -0.206010  0.502292   
...          ...       ...       ...       ...       ...       ...       ...   
284802  1.475829  0.213454  0.111864  1.014480 -0.509348  1.436807  0.250034   
284803  0.059616  0.214205  0.924384  0.012463 -1.016226 -0.606624 -0.395255   
284804  0.001396  0.232045  0.578229 -0.037501  0.640134  0.265745 -0.087371   
284805  0.127434  0.265245  0.800049 -0.163298  0.123205 -0.569159  0.546668   
284806  0.382948  0.261057  0.643078  0.376777  0.008797 -0.473649 -0.818267   

             V27       V28  Amount  
0       0.133558 -0.021053  149.62  
1      -0.008983  0.014724    2.69  
2      -0.055353 -0.059752  378.66  
3       0.062723  0.061458  123.50  
4       0.219422  0.215153   69.99  
...          ...       ...     ...  
28480

In [22]:
y = df['Class']
y, type(y)

(0         0
 1         0
 2         0
 3         0
 4         0
          ..
 284802    0
 284803    0
 284804    0
 284805    0
 284806    0
 Name: Class, Length: 284807, dtype: int64,
 pandas.core.series.Series)

In [23]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=100, stratify=y)

In [24]:
X_train

Time        V1        V2        V3        V4        V5        V6  \
222925  143180.0  2.053474 -0.149880 -1.092234  0.420978 -0.225796 -1.159629   
163811  116231.0  0.149477  1.112008 -0.391539 -0.557452  1.014727 -0.801985   
196038  131332.0  2.145379  0.215237 -2.221643 -0.067032  0.815822 -1.358596   
25173    33527.0 -0.354219  0.100332  1.721509 -0.884417 -0.371977 -0.062313   
253792  156421.0 -0.439888  1.871311 -1.106687  1.064853  1.239594 -0.633668   
...          ...       ...       ...       ...       ...       ...       ...   
194533  130633.0 -0.917471  0.855403  1.516046 -0.220692  0.747332  0.859372   
163532  116011.0 -0.818922 -0.653837  1.685615 -1.561367  0.357487  1.306132   
47169    43114.0  0.984295 -0.951871  0.760848  0.729305 -1.429835 -0.353366   
138367   82618.0  0.699525 -1.558081  0.286748  0.349547 -1.519666 -0.535495   
271        192.0 -0.351287  1.183754  0.530794  0.543891  0.798441 -0.804472   

              V7        V8        V9       V10       V11       V12       V13  \
222925  0.082871 -0.236532  0.733270  0.102520 -0.908262 -0.220082 -1.216003   
163811  1.042734 -0.167625 -0.200830 -0.863839 -0.537991  0.418586  1.034272   
196038  0.843943 -0.594243  0.009745 -0.021497 -0.476749  0.848265  1.144607   
25173  -0.144740  0.163123 -1.843755  0.520706  1.627755 -0.263976 -0.100123   
253792  1.100319  0.221096 -0.917002 -0.844362 -0.888776 -0.219374 -0.076494   
...          ...       ...       ...       ...       ...       ...       ...   
194533  0.827820  0.281375  0.400861 -1.066695 -2.440113 -0.765055 -1.608691   
163532  0.371795 -0.086070 -0.349493  0.156626 -0.675390 -1.141582 -0.944494   
47169  -0.542588 -0.063028 -0.419503  0.629828 -0.971147 -0.123702  0.026550   
138367 -0.117706 -0.177233 -0.468385  0.452339 -0.967100 -0.115463 -0.277494   
271     1.563170 -0.935794  0.633634  2.303627  1.398891 -0.318657 -1.410645   

             V14       V15       V16       V17       V18       V19       V20  \
222925  0.534920  0.227331 -0.103995 -0.158785 -0.802113  0.083010 -0.313465   
163811 -1.222412 -0.349818  0.290628  0.298838 -0.333729 -0.226373  0.084311   
196038  0.555759  0.285353 -0.661516 -0.362293 -0.647254  0.126100 -0.133286   
25173   0.125968  0.703573  0.591865  0.665795 -1.191246  1.690163  0.295444   
253792 -0.756161 -0.170227 -0.554555  1.442972  0.385060  0.585502  0.024939   
...          ...       ...       ...       ...       ...       ...       ...   
194533 -0.233655 -1.496631  0.011731 -0.457877 -0.414743 -0.072057 -0.274203   
163532 -0.998166 -0.265553 -0.894172  1.536807 -2.702692  0.337977  0.072708   
47169  -0.172383  0.361320 -1.416378 -0.041645  1.487895 -1.163956 -0.233008   
138367 -0.071079 -0.190450 -1.633818  0.319327  0.736834 -0.609816  0.116917   
271    -0.155225  0.277950 -0.706228 -0.780092  0.354094  0.618678  0.596401   

             V21       V22       V23       V24       V25       V26       V27  \
222925 -0.298362 -0.779702  0.362701 -0.048183 -0.356438  0.205392 -0.075806   
163811 -0.331768 -0.755443  0.079155  0.587098 -0.383245  0.110292  0.229167   
196038  0.153606  0.611995 -0.027494  0.824538  0.505014  0.203037 -0.083225   
25173   0.136969  0.217167 -0.014350  0.016538 -0.326120 -0.285113  0.118226   
253792  0.059066  0.366613 -0.250861  0.494710  0.091257 -0.398552  0.330173   
...          ...       ...       ...       ...       ...       ...       ...   
194533 -0.583911 -1.674141 -0.248766 -0.213942  0.609341 -1.054254 -0.068467   
163532  0.248178  1.356247 -0.603728 -1.037884  0.609915  0.318683 -0.277412   
47169  -0.148347 -0.163503 -0.190719  0.378213  0.456626 -0.239030  0.045741   
138367 -0.456747 -1.458638 -0.134677  0.368357  0.086125  0.273679 -0.065344   
271    -0.121755  0.668124 -0.057380  0.301398 -0.730946 -0.538616 -0.037212   

             V28  Amount  
222925 -0.063325    1.29  
163811  0.086294    2.69  
196038 -0.069468    4.55  
25173   0.103791   19.95  
253792  0.219909    6.30  
...   

In [25]:
X_test

Time        V1        V2        V3        V4        V5        V6  \
262922  160713.0 -0.548103  1.376074  0.051147  4.300577  1.761938  0.081805   
123418   76919.0  0.437324  1.824381 -2.299325  1.538479  0.791887 -1.414298   
206135  136082.0  2.056846  0.042030 -1.296059  0.368958  0.032669 -1.349797   
273483  165598.0 -0.497480  0.785269  1.695245 -0.644791  0.491553  0.430883   
245758  152894.0  1.611949 -0.938343 -1.200421  0.653356 -0.386518 -0.465937   
...          ...       ...       ...       ...       ...       ...       ...   
109541   71427.0  1.273667  0.348576  0.155665  0.610139 -0.179551 -0.860330   
41813    40839.0 -0.442064 -1.217388  1.414466 -1.427350  0.945895  4.810729   
16722    28090.0  0.933553 -0.288416  0.398369  0.538503  0.280716  1.657391   
36147    38413.0 -4.099698 -3.287632  1.648931  1.793736  1.698853 -0.483157   
198375  132464.0 -7.161413  5.244866 -7.056295  0.519916  3.918486 -2.482685   

               V7         V8        V9       V10       V11       V12  \
262922   1.788166  -0.302164 -2.941339  1.393164 -1.873436 -1.746983   
123418   0.482083   0.339477 -0.674650 -1.234870  1.881042 -0.195098   
206135   0.334441  -0.463096  0.452703 -0.016376 -0.526324  0.801353   
273483   0.510721   0.053386  0.552390 -0.545978 -1.545755 -0.957582   
245758   0.069881  -0.147945  0.849526  0.005008 -1.033421 -0.111964   
...           ...        ...       ...       ...       ...       ...   
109541   0.100726  -0.195347  0.020811 -0.291052 -0.347286  0.242342   
41813   -1.657637   1.121731  0.529958  0.344311 -1.165678 -0.613130   
16722   -0.408491   0.596638  0.174499 -0.234109  2.003246  1.568237   
36147   -0.603089   0.271586  0.127974  0.486426  1.213868  0.653035   
198375 -14.996357 -15.904451 -1.835893 -4.195324 -0.599012  2.849676   

             V13       V14       V15       V16       V17       V18       V19  \
262922 -0.852186  1.058096  0.371916 -0.783216  0.322476  0.210874  1.701892   
123418 -0.688822 -2.926057  0.427144  1.279861  2.651797  2.224899 -0.000361   
206135  1.006171  0.303048  0.637776 -0.262217 -0.553146 -0.181507 -0.244297   
273483 -0.913807 -0.066661  1.561068  0.043679 -0.672767  0.650035  1.013564   
245758 -0.645005  0.305443  0.414153  0.183540 -0.410625 -0.207679 -0.171987   
...          ...       ...       ...       ...       ...       ...       ...   
109541  0.526277 -0.355216  1.094805  0.572843 -0.149382 -0.210053 -0.065830   
41813  -0.028020 -1.823057 -1.339567  0.385304  0.595942 -0.497526  2.165032   
16722  -0.135354  0.244879  0.175965 -1.350302  0.983088 -2.162239 -1.054337   
36147   0.600744 -0.210172  1.429186  0.531294 -0.727684  0.778147  0.518246   
198375 -3.349200  3.432775 -1.566190  1.768522  3.985681  1.689386 -1.590626   

             V20       V21       V22       V23       V24       V25       V26  \
262922  0.451713  0.316647  0.700714 -0.532490 -0.680250  1.232858  0.980223   
123418 -0.009730 -0.077049 -0.170282  0.108469 -0.365513 -0.436975 -0.393859   
206135 -0.167702  0.261879  0.949829 -0.018148  0.109873  0.295611 -0.115295   
273483  0.212146 -0.280792 -0.548862 -0.367317 -0.044278  0.430846 -0.657662   
245758  0.214406  0.081319 -0.184030  0.065733  0.742927 -0.283315  0.172087   
...          ...       ...       ...       ...       ...       ...       ...   
109541 -0.045382 -0.305978 -0.878975  0.078359 -0.008235  0.270545  0.115470   
41813   0.391646  0.309636  1.298415 -0.227884  1.109008 -0.265818  0.066769   
16722  -0.169897 -0.038509  0.109277  0.095344 -0.983303  0.097938  0.376637   
36147  -1.029740 -0.219962  1.209664  2.771741 -0.317573  0.388055 -0.137592   
198375  2.685353 -7.997422  2.642809 -8.535706 -0.170266 -3.930720 -1.005978   

             V27       V28  Amount  
262922 -0.042880  0.053000  101.42  
123418  0.121535 -0.061320    9.99  
206135 -0.015115 -0.054976   10.49  
273483  0.179848 -0.081837    1.00  
245758 -0.091352 -0.009690  227.75  
...          ...       ...     ...  
10954

In [26]:
y_train

222925    0
163811    0
196038    0
25173     0
253792    0
         ..
194533    0
163532    0
47169     0
138367    0
271       0
Name: Class, Length: 199364, dtype: int64

In [27]:
y_test

262922    0
123418    0
206135    0
273483    0
245758    0
         ..
109541    0
41813     0
16722     0
36147     0
198375    0
Name: Class, Length: 85443, dtype: int64

In [28]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 199364 entries, 222925 to 271
Data columns (total 30 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    199364 non-null  float64
 1   V1      199364 non-null  float64
 2   V2      199364 non-null  float64
 3   V3      199364 non-null  float64
 4   V4      199364 non-null  float64
 5   V5      199364 non-null  float64
 6   V6      199364 non-null  float64
 7   V7      199364 non-null  float64
 8   V8      199364 non-null  float64
 9   V9      199364 non-null  float64
 10  V10     199364 non-null  float64
 11  V11     199364 non-null  float64
 12  V12     199364 non-null  float64
 13  V13     199364 non-null  float64
 14  V14     199364 non-null  float64
 15  V15     199364 non-null  float64
 16  V16     199364 non-null  float64
 17  V17     199364 non-null  float64
 18  V18     199364 non-null  float64
 19  V19     199364 non-null  float64
 20  V20     199364 non-null  float64
 21  V21     

In [29]:
X_train.describe()

Time             V1             V2             V3  \
count  199364.000000  199364.000000  199364.000000  199364.000000   
mean    94810.339675       0.002243      -0.001196       0.000810   
std     47474.327394       1.957553       1.652392       1.512832   
min         0.000000     -56.407510     -72.715728     -32.965346   
25%     54173.250000      -0.917434      -0.598699      -0.888159   
50%     84654.500000       0.020646       0.064743       0.181180   
75%    139366.250000       1.314864       0.801737       1.029009   
max    172792.000000       2.454930      22.057729       4.226108   

                  V4             V5             V6             V7  \
count  199364.000000  199364.000000  199364.000000  199364.000000   
mean        0.001849       0.002252       0.000860      -0.002396   
std         1.417121       1.365517       1.325449       1.218740   
min        -5.683171     -42.147898     -26.160506     -43.557242   
25%        -0.849660      -0.691944      -0.767172      -0.554808   
50%        -0.019098      -0.055280      -0.271647       0.039937   
75%         0.746461       0.612686       0.402208       0.570538   
max        16.715537      34.801666      23.917837      44.054461   

                  V8             V9            V10            V11  \
count  199364.000000  199364.000000  199364.000000  199364.000000   
mean       -0.002061      -0.001636      -0.002549      -0.001743   
std         1.208945       1.096996       1.082398       1.018143   
min       -73.216718     -13.434066     -24.588262      -4.568390   
25%        -0.208512      -0.644613      -0.537343      -0.762321   
50%         0.022900      -0.051257      -0.093973      -0.034723   
75%         0.328388       0.596602       0.453867       0.737656   
max        20.007208      10.392889      15.331742      12.018913   

                 V12            V13            V14            V15  \
count  199364.000000  199364.000000  199364.000000  199364.000000   
mean        0.001404      -0.001305       0.000040       0.000068   
std         0.994797       0.995828       0.955804       0.915469   
min       -18.683715      -5.791881     -19.214325      -4.498945   
25%        -0.404682      -0.650200      -0.425083      -0.582549   
50%         0.139429      -0.015088       0.050493       0.048497   
75%         0.619286       0.662928       0.491942       0.649400   
max         7.848392       4.465413      10.526766       5.825654   

                 V16            V17            V18            V19  \
count  199364.000000  199364.000000  199364.000000  199364.000000   
mean        0.000483       0.001630       0.000303      -0.000020   
std         0.871720       0.841630       0.838102       0.814197   
min       -14.129855     -23.815636      -9.498746      -7.213527   
25%        -0.468477      -0.483477      -0.498889      -0.457060   
50%         0.065720      -0.064986      -0.002910       0.004455   
75%         0.522746       0.401246       0.501645       0.457607   
max         6.442798       9.253526       5.041069       5.591971   

                 V20            V21            V22            V23  \
count  199364.000000  199364.000000  199364.000000  199364.000000   
mean       -0.000126      -0.000035      -0.000532      -0.001047   
std         0.760362       0.746103       0.726956       0.625258   
min       -25.222345     -34.830382     -10.933144     -44.807735   
25%        -0.211651      -0.228272      -0.542095      -0.162000   
50%        -0.062606      -0.029201       0.006807      -0.011571   
75%         0.133509       0.186293       0.527498       0.147256   
max        39.420904      27.202839      10.503090      22.528412   

                 V24            V25            V26            V27  \
count  199364.000000  199364.000000  199364.000000  199364.000000   
mean       -0.000690       0.000685      -0.000254       0.000120   
std         0.606662       0.521484       0.481503       0.396735   
min     

In [30]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 85443 entries, 262922 to 198375
Data columns (total 30 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Time    85443 non-null  float64
 1   V1      85443 non-null  float64
 2   V2      85443 non-null  float64
 3   V3      85443 non-null  float64
 4   V4      85443 non-null  float64
 5   V5      85443 non-null  float64
 6   V6      85443 non-null  float64
 7   V7      85443 non-null  float64
 8   V8      85443 non-null  float64
 9   V9      85443 non-null  float64
 10  V10     85443 non-null  float64
 11  V11     85443 non-null  float64
 12  V12     85443 non-null  float64
 13  V13     85443 non-null  float64
 14  V14     85443 non-null  float64
 15  V15     85443 non-null  float64
 16  V16     85443 non-null  float64
 17  V17     85443 non-null  float64
 18  V18     85443 non-null  float64
 19  V19     85443 non-null  float64
 20  V20     85443 non-null  float64
 21  V21     85443 non-null  float

In [31]:
X_test.describe()

Time            V1            V2            V3            V4  \
count   85443.000000  85443.000000  85443.000000  85443.000000  85443.000000   
mean    94822.072551     -0.005233      0.002791     -0.001889     -0.004314   
std     47520.650059      1.961361      1.648783      1.524219      1.412940   
min         1.000000    -37.558067    -60.464618    -48.325589     -5.519697   
25%     54269.500000     -0.926877     -0.598137     -0.894314     -0.845380   
50%     84778.000000      0.011386      0.067328      0.176664     -0.021755   
75%    139221.500000      1.318048      0.808287      1.023189      0.736470   
max    172788.000000      2.446505     18.902453      9.382558     16.875344   

                 V5            V6            V7            V8            V9  \
count  85443.000000  85443.000000  85443.000000  85443.000000  85443.000000   
mean      -0.005255     -0.002008      0.005591      0.004808      0.003817   
std        1.414013      1.348061      1.278883      1.159585      1.102436   
min     -113.743307    -20.869626    -31.197329    -50.688419     -9.462573   
25%       -0.690840     -0.770998     -0.551829     -0.208711     -0.639604   
50%       -0.052732     -0.279775      0.040358      0.020809     -0.051858   
75%        0.609618      0.389522      0.570271      0.324674      0.598522   
max       28.516513     73.301626    120.589494     18.329406     15.594995   

                V10           V11           V12           V13           V14  \
count  85443.000000  85443.000000  85443.000000  85443.000000  85443.000000   
mean       0.005948      0.004066     -0.003276      0.003044     -0.000093   
std        1.103741      1.026679      1.009401      0.993979      0.965083   
min      -22.187089     -4.797473    -18.553697     -3.888606    -18.392091   
25%       -0.529999     -0.762999     -0.407139     -0.644664     -0.426373   
50%       -0.090538     -0.028257      0.141871     -0.010485      0.050899   
75%        0.454055      0.745123      0.616102      0.661400      0.495790   
max       23.745136     11.152491      4.574082      7.126883      7.667726   

                V15           V16           V17           V18           V19  \
count  85443.000000  85443.000000  85443.000000  85443.000000  85443.000000   
mean      -0.000159     -0.001126     -0.003803     -0.000708      0.000046   
std        0.914964      0.886743      0.867047      0.838353      0.813680   
min       -4.152532    -13.256833    -25.162799     -9.335193     -6.938297   
25%       -0.583488     -0.467478     -0.484610     -0.497563     -0.454650   
50%        0.046613      0.067906     -0.067270     -0.004873      0.002128   
75%        0.647365      0.524394      0.396579      0.498310      0.461638   
max        8.877742     17.315112      7.733456      4.712398      5.501747   

                V20           V21           V22           V23           V24  \
count  85443.000000  85443.000000  85443.000000  85443.000000  85443.000000   
mean       0.000294      0.000081      0.001241      0.002443      0.001610   
std        0.795030      0.706774      0.722769      0.622593      0.603274   
min      -54.497720    -22.757540     -9.499423    -36.666000     -2.789614   
25%       -0.211894     -0.228753     -0.542937     -0.161423     -0.352905   
50%       -0.062060     -0.029883      0.006755     -0.010201      0.042011   
75%        0.132264      0.186560      0.531061      0.148458      0.439078   
max       38.117209     27.202839      7.357255     22.083545      4.584549   

                V25           V26           V27           V28        Amount  
count  85443.000000  85443.000000  85443.000000  85443.000000  85443.000000  
mean      -0.001598      0.000592     -0.000280     -0.000732     88.839198  
std        0.520798      0.483914      0.419287      0.332254    267.137085  
min       -7.081325     -2.604551    -22.565679    -15.430084      0.000000  
25%       -0.318637     -0.327328     -0.071009     -0.05

In [32]:
y_train.describe()

count    199364.000000
mean          0.001725
std           0.041503
min           0.000000
25%           0.000000
50%           0.000000
75%           0.000000
max           1.000000
Name: Class, dtype: float64

In [33]:
y_test.describe()

count    85443.000000
mean         0.001732
std          0.041583
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max          1.000000
Name: Class, dtype: float64

In [34]:
parameters = [{'n_estimators': [10, 15], 'max_features': np.arange(3, 5), 'max_depth': np.arange(4, 7)}]

In [35]:
model =  GridSearchCV(estimator=RandomForestClassifier(random_state=100), param_grid=parameters, scoring='roc_auc', cv=3)

In [36]:
model.fit(X_train, y_train)

GridSearchCV(cv=3, estimator=RandomForestClassifier(random_state=100),
             param_grid=[{'max_depth': array([4, 5, 6]),
                          'max_features': array([3, 4]),
                          'n_estimators': [10, 15]}],
             scoring='roc_auc')

In [37]:
model.best_params_

{'max_depth': 6, 'max_features': 3, 'n_estimators': 15}

In [38]:
y_pred_proba = model.predict_proba(X_test)[:,1]
y_pred_proba

array([0.00092917, 0.00029521, 0.00028215, ..., 0.00028215, 0.0006822 ,
       0.01246098])

In [39]:
from sklearn.metrics import roc_auc_score

In [40]:
roc_auc_score(y_test, y_pred_proba)

0.9462664156037156

### *Дополнительные задания:

1). Загрузите датасет Wine из встроенных датасетов sklearn.datasets с помощью функции load_wine в переменную data.

In [41]:
from sklearn.datasets import load_wine

In [42]:
data = load_wine()

2). Полученный датасет не является датафреймом. Это структура данных, имеющая ключи аналогично словарю. Просмотрите тип данных этой структуры данных и создайте список data_keys, содержащий ее ключи.

In [43]:
data_keys = data.keys()
type(data), data_keys

(sklearn.utils.Bunch,
 dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names']))

3). Просмотрите данные, описание и названия признаков в датасете. Описание нужно вывести в виде привычного, аккуратно оформленного текста, без обозначений переноса строки, но с самими переносами и т.д.

In [44]:
data.data

array([[1.423e+01, 1.710e+00, 2.430e+00, ..., 1.040e+00, 3.920e+00,
        1.065e+03],
       [1.320e+01, 1.780e+00, 2.140e+00, ..., 1.050e+00, 3.400e+00,
        1.050e+03],
       [1.316e+01, 2.360e+00, 2.670e+00, ..., 1.030e+00, 3.170e+00,
        1.185e+03],
       ...,
       [1.327e+01, 4.280e+00, 2.260e+00, ..., 5.900e-01, 1.560e+00,
        8.350e+02],
       [1.317e+01, 2.590e+00, 2.370e+00, ..., 6.000e-01, 1.620e+00,
        8.400e+02],
       [1.413e+01, 4.100e+00, 2.740e+00, ..., 6.100e-01, 1.600e+00,
        5.600e+02]])

In [45]:
data.data.shape

(178, 13)

In [46]:
data.feature_names

['alcohol',
 'malic_acid',
 'ash',
 'alcalinity_of_ash',
 'magnesium',
 'total_phenols',
 'flavanoids',
 'nonflavanoid_phenols',
 'proanthocyanins',
 'color_intensity',
 'hue',
 'od280/od315_of_diluted_wines',
 'proline']

In [47]:
print(data.DESCR)

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

4). Сколько классов содержит целевая переменная датасета? Выведите названия классов.

In [48]:
data.target_names

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

In [49]:
np.unique(data.target, return_counts=True)

(array([0, 1, 2]), array([59, 71, 48], dtype=int64))

Отсюда делаем вывод, что целеавая переменная имеет 3 класса: class_0, class_1, class_2

5). На основе данных датасета (они содержатся в двумерном массиве Numpy) и названий признаков создайте датафрейм под названием X.

In [50]:
X = pd.DataFrame(data.data, columns = data.feature_names)
X

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  
0                            3.92   1065.0  
1                            3.40   1050.0  
2                            3.17   1185.0  
3                            3.45   1480.0  
4                            2.93    735.0  
..                            ...      ...  
173                          1.74    740.0  
174                          1.56    750.0  
175                          1.56    835.0  
176                          1.62    840.0  
177                          1.60    560.0  

[178 rows x 13 columns]

6). Выясните размер датафрейма X и установите, имеются ли в нем пропущенные значения.

In [51]:
X.shape

(178, 13)

In [52]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 13 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   alcohol                       178 non-null    float64
 1   malic_acid                    178 non-null    float64
 2   ash                           178 non-null    float64
 3   alcalinity_of_ash             178 non-null    float64
 4   magnesium                     178 non-null    float64
 5   total_phenols                 178 non-null    float64
 6   flavanoids                    178 non-null    float64
 7   nonflavanoid_phenols          178 non-null    float64
 8   proanthocyanins               178 non-null    float64
 9   color_intensity               178 non-null    float64
 10  hue                           178 non-null    float64
 11  od280/od315_of_diluted_wines  178 non-null    float64
 12  proline                       178 non-null    float64
dtypes: fl

7). Добавьте в датафрейм поле с классами вин в виде чисел, имеющих тип данных numpy.int64. Название поля - 'target'.

In [53]:
X['target'] = data.target.astype(np.int64)
X

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  target  
0                            3.92   1065.0       0  
1                            3.40   1050.0       0  
2                            3.17   1185.0       0  
3                            3.45   1480.0       0  
4                            2.93    735.0       0  
..                            ...      ...     ...  
173                          1.74    740.0       2  
174                          1.56    750.0       2  
175                          1.56    835.0       2  
176                          1.62    840.0       2  
177                          1.60    560.0       2  

[178 rows x 14 columns]

In [54]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 14 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   alcohol                       178 non-null    float64
 1   malic_acid                    178 non-null    float64
 2   ash                           178 non-null    float64
 3   alcalinity_of_ash             178 non-null    float64
 4   magnesium                     178 non-null    float64
 5   total_phenols                 178 non-null    float64
 6   flavanoids                    178 non-null    float64
 7   nonflavanoid_phenols          178 non-null    float64
 8   proanthocyanins               178 non-null    float64
 9   color_intensity               178 non-null    float64
 10  hue                           178 non-null    float64
 11  od280/od315_of_diluted_wines  178 non-null    float64
 12  proline                       178 non-null    float64
 13  targe

8). Постройте матрицу корреляций для всех полей X. Дайте полученному датафрейму название X_corr.

In [55]:
X_corr = X.corr()
X_corr

alcohol  malic_acid       ash  \
alcohol                       1.000000    0.094397  0.211545   
malic_acid                    0.094397    1.000000  0.164045   
ash                           0.211545    0.164045  1.000000   
alcalinity_of_ash            -0.310235    0.288500  0.443367   
magnesium                     0.270798   -0.054575  0.286587   
total_phenols                 0.289101   -0.335167  0.128980   
flavanoids                    0.236815   -0.411007  0.115077   
nonflavanoid_phenols         -0.155929    0.292977  0.186230   
proanthocyanins               0.136698   -0.220746  0.009652   
color_intensity               0.546364    0.248985  0.258887   
hue                          -0.071747   -0.561296 -0.074667   
od280/od315_of_diluted_wines  0.072343   -0.368710  0.003911   
proline                       0.643720   -0.192011  0.223626   
target                       -0.328222    0.437776 -0.049643   

                              alcalinity_of_ash  magnesium  total_phenols  \
alcohol                               -0.310235   0.270798       0.289101   
malic_acid                             0.288500  -0.054575      -0.335167   
ash                                    0.443367   0.286587       0.128980   
alcalinity_of_ash                      1.000000  -0.083333      -0.321113   
magnesium                             -0.083333   1.000000       0.214401   
total_phenols                         -0.321113   0.214401       1.000000   
flavanoids                            -0.351370   0.195784       0.864564   
nonflavanoid_phenols                   0.361922  -0.256294      -0.449935   
proanthocyanins                       -0.197327   0.236441       0.612413   
color_intensity                        0.018732   0.199950      -0.055136   
hue                                   -0.273955   0.055398       0.433681   
od280/od315_of_diluted_wines          -0.276769   0.066004       0.699949   
proline                               -0.440597   0.393351       0.498115   
target                                 0.517859  -0.209179      -0.719163   

                              flavanoids  nonflavanoid_phenols  \
alcohol                         0.236815             -0.155929   
malic_acid                     -0.411007              0.292977   
ash                             0.115077              0.186230   
alcalinity_of_ash              -0.351370              0.361922   
magnesium                       0.195784             -0.256294   
total_phenols                   0.864564             -0.449935   
flavanoids                      1.000000             -0.537900   
nonflavanoid_phenols           -0.537900              1.000000   
proanthocyanins                 0.652692             -0.365845   
color_intensity                -0.172379              0.139057   
hue                             0.543479             -0.262640   
od280/od315_of_diluted_wines    0.787194             -0.503270   
proline                         0.494193             -0.311385   
target                         -0.847498              0.489109   

                              proanthocyanins  color_intensity       hue  \
alcohol                              0.136698         0.546364 -0.071747   
malic_acid                          -0.220746         0.248985 -0.561296   
ash                                  0.009652         0.258887 -0.074667   
alcalinity_of_ash                   -0.197327         0.018732 -0.273955   
magnesium                            0.236441         0.199950  0.055398   
total_phenols                        0.612413        -0.055136  0.433681   
flavanoids                           0.652692        -0.172379  0.543479   
nonflavanoid_phenols                -0.365845         0.139057 -0.262640   
proanthocyanins                      1.000000        -0.025250  0.295544   
color_intensity                     -0.025250         1.000000 -0.521813   
hue                                  0.295544        -0.521813  1.000000   
od280/od315

9). Создайте список high_corr из признаков, корреляция которых с полем target по абсолютному значению превышает 0.5 (причем, само поле target не должно входить в этот список).

In [56]:
high_corr = X_corr[np.abs(X_corr['target'])>0.5].drop('target')
high_corr

alcohol  malic_acid       ash  \
alcalinity_of_ash            -0.310235    0.288500  0.443367   
total_phenols                 0.289101   -0.335167  0.128980   
flavanoids                    0.236815   -0.411007  0.115077   
hue                          -0.071747   -0.561296 -0.074667   
od280/od315_of_diluted_wines  0.072343   -0.368710  0.003911   
proline                       0.643720   -0.192011  0.223626   

                              alcalinity_of_ash  magnesium  total_phenols  \
alcalinity_of_ash                      1.000000  -0.083333      -0.321113   
total_phenols                         -0.321113   0.214401       1.000000   
flavanoids                            -0.351370   0.195784       0.864564   
hue                                   -0.273955   0.055398       0.433681   
od280/od315_of_diluted_wines          -0.276769   0.066004       0.699949   
proline                               -0.440597   0.393351       0.498115   

                              flavanoids  nonflavanoid_phenols  \
alcalinity_of_ash              -0.351370              0.361922   
total_phenols                   0.864564             -0.449935   
flavanoids                      1.000000             -0.537900   
hue                             0.543479             -0.262640   
od280/od315_of_diluted_wines    0.787194             -0.503270   
proline                         0.494193             -0.311385   

                              proanthocyanins  color_intensity       hue  \
alcalinity_of_ash                   -0.197327         0.018732 -0.273955   
total_phenols                        0.612413        -0.055136  0.433681   
flavanoids                           0.652692        -0.172379  0.543479   
hue                                  0.295544        -0.521813  1.000000   
od280/od315_of_diluted_wines         0.519067        -0.428815  0.565468   
proline                              0.330417         0.316100  0.236183   

                              od280/od315_of_diluted_wines   proline    target  
alcalinity_of_ash                                -0.276769 -0.440597  0.517859  
total_phenols                                     0.699949  0.498115 -0.719163  
flavanoids                                        0.787194  0.494193 -0.847498  
hue                                               0.565468  0.236183 -0.617369  
od280/od315_of_diluted_wines                      1.000000  0.312761 -0.788230  
proline                                           0.312761  1.000000 -0.633717

10). Удалите из датафрейма X поле с целевой переменной. Для всех признаков, названия которых содержатся в списке high_corr, вычислите квадрат их значений и добавьте в датафрейм X соответствующие поля с суффиксом '_2', добавленного к первоначальному названию признака. Итоговый датафрейм должен содержать все поля, которые, были в нем изначально, а также поля с признаками из списка high_corr, возведенными в квадрат. Выведите описание полей датафрейма X с помощью метода describe.

In [57]:
X = X.drop(columns = ['target'])

In [58]:
X[high_corr.index.values + '_2'] = np.power(X[high_corr.index.values], 2)
X

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  alcalinity_of_ash_2  \
0                            3.92   1065.0               243.36   
1                            3.40   1050.0               125.44   
2                            3.17   1185.0               345.96   
3                            3.45   1480.0               282.24   
4                            2.93    735.0               441.00   
..                            ...      ...                  ...   
173                          1.74    740.0               420.25   
174                          1.56    750.0               529.00   
175                          1.56    835.0               400.00   
176                          1.62    840.0               400.00   
177                          1.60    560.0               600.25   

     total_phenols_2  flavanoids_2   hue_2  od280/od315_of_diluted_wines_2  \
0             7.8400        9.3636  1.0816                         15.3664   
1             7.0225        7.6176  1.1025                         11.5600   
2             7.8400       10.4976  1.0609                         10.0489   
3            14.8225       12.1801  0.7396                         11.9025   
4             7.8400        7.2361  1.0816                          8.5849   
..               ...           ...     ...                             ...   
173           2.8224        0.3721  0.4096                          3.0276   
174           3.2400        0.5625  0.4900                          2.4336   
175           2.5281        0.4761  0.3481                          2.4336   
176           2.7225        0.4624  0.3600                          2.6244   
177           4.2025        0.5776  0.3721                          2.5600   

     proline_2  
0    1134225.0  
1    1102500.0  
2    1404225.0  
3    2190400.0  
4     540225.0  
..         ...  
173   547600.0  
174   562500.0  
175   697225.0  
176   705600.0  
177   313600.0  

[178 rows x 19 columns]

In [59]:
X.describe()

alcohol  malic_acid         ash  alcalinity_of_ash   magnesium  \
count  178.000000  178.000000  178.000000         178.000000  178.000000   
mean    13.000618    2.336348    2.366517          19.494944   99.741573   
std      0.811827    1.117146    0.274344           3.339564   14.282484   
min     11.030000    0.740000    1.360000          10.600000   70.000000   
25%     12.362500    1.602500    2.210000          17.200000   88.000000   
50%     13.050000    1.865000    2.360000          19.500000   98.000000   
75%     13.677500    3.082500    2.557500          21.500000  107.000000   
max     14.830000    5.800000    3.230000          30.000000  162.000000   

       total_phenols  flavanoids  nonflavanoid_phenols  proanthocyanins  \
count     178.000000  178.000000            178.000000       178.000000   
mean        2.295112    2.029270              0.361854         1.590899   
std         0.625851    0.998859              0.124453         0.572359   
min         0.980000    0.340000              0.130000         0.410000   
25%         1.742500    1.205000              0.270000         1.250000   
50%         2.355000    2.135000              0.340000         1.555000   
75%         2.800000    2.875000              0.437500         1.950000   
max         3.880000    5.080000              0.660000         3.580000   

       color_intensity         hue  od280/od315_of_diluted_wines      proline  \
count       178.000000  178.000000                    178.000000   178.000000   
mean          5.058090    0.957449                      2.611685   746.893258   
std           2.318286    0.228572                      0.709990   314.907474   
min           1.280000    0.480000                      1.270000   278.000000   
25%           3.220000    0.782500                      1.937500   500.500000   
50%           4.690000    0.965000                      2.780000   673.500000   
75%           6.200000    1.120000                      3.170000   985.000000   
max          13.000000    1.710000                      4.000000  1680.000000   

       alcalinity_of_ash_2  total_phenols_2  flavanoids_2       hue_2  \
count           178.000000       178.000000    178.000000  178.000000   
mean            391.142865         5.657030      5.110049    0.968661   
std             133.671775         2.936294      4.211441    0.443798   
min             112.360000         0.960400      0.115600    0.230400   
25%             295.840000         3.036325      1.452100    0.612325   
50%             380.250000         5.546050      4.558250    0.931250   
75%             462.250000         7.840000      8.265700    1.254400   
max             900.000000        15.054400     25.806400    2.924100   

       od280/od315_of_diluted_wines_2     proline_2  
count                      178.000000  1.780000e+02  
mean                         7.322155  6.564591e+05  
std                          3.584316  5.558591e+05  
min                          1.612900  7.728400e+04  
25%                          3.754075  2.505010e+05  
50%                          7.728400  4.536045e+05  
75%                         10.048900  9.702250e+05  
max                         16.000000  2.822400e+06

In [60]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 19 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   alcohol                         178 non-null    float64
 1   malic_acid                      178 non-null    float64
 2   ash                             178 non-null    float64
 3   alcalinity_of_ash               178 non-null    float64
 4   magnesium                       178 non-null    float64
 5   total_phenols                   178 non-null    float64
 6   flavanoids                      178 non-null    float64
 7   nonflavanoid_phenols            178 non-null    float64
 8   proanthocyanins                 178 non-null    float64
 9   color_intensity                 178 non-null    float64
 10  hue                             178 non-null    float64
 11  od280/od315_of_diluted_wines    178 non-null    float64
 12  proline                         178 